File: uva/Factovisors/Factorvisors.cpp
#include <iostream>
#include <map>
#include <vector>
typedef long long ll;
using namespace std;

int main(){
    double n,m;

    map<double,double>mp;
    double max = 1024;
    double temp = 1;
    for(double i = 1; i <= max; i++){
        mp[i] = temp;
        temp *= (i+1);
    }

    while(cin >> n >> m){
        if(m == 0){
            printf("0 divides %.0f!\n", n);
            continue;
        }

        if(m < n){
            printf("%.0f divides %.0f!\n",m, n);
            continue;
        }

        ll resto = ll(m) % ll(mp[n]);
        if(m == 1 || resto == 0){
            printf("%.0f divides %.0f!\n",m, n);
        }else{
            printf("%.0f does not divide %.0f!\n", m, n);
        }
    }
}

File: uva/Factovisors/Factorvisors.java
import java.math.BigDecimal;
import java.util.Scanner;

public class Factorvisors{
    public static void main(String args[]){
        Scanner sc = new Scanner(System.in);
        BigDecimal n,m;

        while(sc.hasNext()){
            n = sc.nextBigDecimal();
            m = sc.nextBigDecimal();

            BigDecimal fatorial = BigDecimal.ONE;
            BigDecimal tempn = n;
            for(int i = 0; i < n.intValue(); i++){
                fatorial = fatorial.multiply(tempn);
                tempn = tempn.subtract( BigDecimal.ONE);
            }

            if(fatorial.remainder(m) == BigDecimal.ZERO){
                System.out.printf("%.0f divides %.0f!\n", m,n);
            }else{
                System.out.printf("%.0f does not divide %.0f!\n", m,n);
            }
        }
        
    }   
}   